gandi/structures_domain.go

```diff
@@ -45,31 +45,42 @@ var flattenContactType = []string{
 }
 
 func expandContact(in interface{}) *domain.Contact {
+	var cnt domain.Contact
 	list := in.(*schema.Set).List()
-	// We are sure the TypeSet contains a single element thanks to
-	// the MaxItems=1 constraint
-	contact := list[0].(map[string]interface{})
 
-	dataObfuscated := contact["data_obfuscated"].(bool)
-	mailObfuscated := contact["mail_obfuscated"].(bool)
+	// For an unknown reason, Terraform provides two TypeSet blocks
+	// on an update: one is empty, while the other one contains
+	// the actual data. This seems to be a known issue :/
+	// See https://discuss.hashicorp.com/t/using-typeset-in-provider-always-adds-an-empty-element-on-update/18566
+	for _, elt := range list {
+		contact := elt.(map[string]interface{})
 
-	cnt := domain.Contact{
-		Country:         contact["country"].(string),
-		State:           contact["state"].(string),
-		DataObfuscated:  &dataObfuscated,
-		MailObfuscated:  &mailObfuscated,
-		Email:           contact["email"].(string),
-		FamilyName:      contact["family_name"].(string),
-		GivenName:       contact["given_name"].(string),
-		StreetAddr:      contact["street_addr"].(string),
-		Phone:           contact["phone"].(string),
-		City:            contact["city"].(string),
-		OrgName:         contact["organisation"].(string),
-		Zip:             contact["zip"].(string),
-		ContactType:     expandContactType[contact["type"].(string)],
-		ExtraParameters: contact["extra_parameters"].(map[string]interface{}),
+		dataObfuscated := contact["data_obfuscated"].(bool)
+		mailObfuscated := contact["mail_obfuscated"].(bool)
+
+		cnt = domain.Contact{
+			Country:         contact["country"].(string),
+			State:           contact["state"].(string),
+			DataObfuscated:  &dataObfuscated,
+			MailObfuscated:  &mailObfuscated,
```

Comment:
The AWS libs do of taking the address of a thing. They have helper functions that are basically:

```
package aws

func Bool(in bool) *bool {
    return &in
}
```

Which then lets you write:

```
Bool(contact["data_obfuscated"].(bool))
```

Here on line 65.

Reply:
Thx for the tips!
However, i would prefer to avoid introducing a dependency just for this small improvement.

Reply:
Of course, I meant more of adding such a function yourself. The AWS library is way way way too big to introduce for such a small thing.

Reply:
Applied.